services/retention/service.go

```diff
@@ -79,7 +79,7 @@ func (s *Service) run() {
 			return
 
 		case <-ticker.C:
-			s.logger.Info("Retention policy shard deletion check commencing.")
+			log, logEnd := logger.NewOperation(s.logger, "Retention policy deletion check", "retention.delete_check")
```

Comment:
This call and the eventual invocation are far enough away that I think this entire section deserves a separate function. Either an inline closure or, since the indentation seems to be becoming too deep, you might want to consider moving this operation into its own function.

I think since the body is this long, a defer is more suitable here.